### PR TITLE
Suggest PHP `ext-ev`

### PR DESCRIPTION
**Pull request by yhabteab**
Currently, the React PHP uses `StreamSelect` eventloop, which limits you to only have 1024 open FDs in parallel. However, since the scheduler can run several jobs from any module, this limit can be reached very quickly and will result in a crash with a fatal error. I was even able to reach this limit only with **one** very long running job. See here for reference https://daniel.haxx.se/docs/poll-vs-select.html

This event can run until the memory is exhausted:
```php
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 16384 bytes) in /usr/local/src/ipl-sql/src/Connection.php on line 401
PHP Stack trace:
PHP   1. React\EventLoop\Loop::React\EventLoop\{closure:/usr/share/icinga-php/vendor/vendor/react/event-loop/src/Loop.php:47-57}() /usr/share/icinga-php/vendor/vendor/react/event-loop/src/Loop.php:0
PHP   2. React\EventLoop\ExtEvLoop->run() /usr/share/icinga-php/vendor/vendor/react/event-loop/src/Loop.php:55
PHP   3. React\EventLoop\Tick\FutureTickQueue->tick() /usr/share/icinga-php/vendor/vendor/react/event-loop/src/ExtEvLoop.php:192
```

### CODE REVIEW

**Comment by nilmerg** [comment]
And why does installing this extension help?

**Comment by yhabteab** [comment]
> And why does installing this extension help?

That’s why. https://github.com/reactphp/event-loop/blob/1.x/src/Factory.php#L47-L73

It automatically tries to select an event loop based on the installed extensions, and if none are installed, it uses Stream Select.

File: composer.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
   "require": {
     "php": ">=7.2",
     "ext-json": "*",
+    "ext-ev": "*",
```
Comment (lippserd) [inline composer.json:14]: Instead of require, I would use [suggest](https://getcomposer.org/doc/04-schema.md#suggest). This should be a separate commit.

File: tests/SchedulerTest.php [hunk 2]
```diff
@@ -114,14 +114,13 @@ function (Task $_, array $promises) use (&$canceledPromises): void {
                 }
             );
 
-        // Tick callbacks are guaranteed to be executed in the order they are enqueued,
-        // thus we will not be able to remove the task before it is finished. Therefore,
-        // we need to use timers with 0 intervals.
-        Loop::addTimer(0, function () use ($task): void {
+        Loop::addTimer(.001, function () use ($task): void {
```
Comment (lippserd) [inline tests/SchedulerTest.php:117]: Please explain, why we're sleeping for `.001` seconds. This change should also be in a separate commit.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: composer.json [hunk 3]
```diff
@@ -19,6 +19,9 @@
     "simshaun/recurr": "^5",
     "ipl/stdlib": ">=0.12.0"
   },
+  "suggest": {
+    "ext-ev": "Allows to have over 1024 open file descriptors in parallel. Additionally, it improves performance with just a handful of FDs, and the benefit grows with each additional connection."
```
Comment (nilmerg) [inline composer.json:23]: > each additional *connection*

The audience of this message are users of this library, right? Can we provide a message then that doesn't explain an implementation detail?

Comment (yhabteab) [reply composer.json:23]: I don't get what is supposed to be implementation details in the quoted words.

Comment (nilmerg) [reply composer.json:23]: This library is used to schedule jobs. How it does that, is an implementation detail. Whether it uses syscalls or it's own timer implementation is irrelevant to the user of it, or not?

Comment (lippserd) [reply composer.json:23]: At least I consider the information about 1024 maximum parallel connections worthwhile.

Comment (nilmerg) [reply composer.json:23]: You should really get into the habit of explaining your reasoning :wink:

Comment (yhabteab) [reply composer.json:23]: Which kind of explanation should we give to the user then, when we don't mention the actual reason for installing it?

Comment (nilmerg) [reply composer.json:23]: *Improves performance, efficiency and avoids system limitations. Highly recommended! (See $link for details)*
